### PR TITLE
Allow filters to be joined using 'Or' operator

### PR DESCRIPTION
**Pull request by E-Paine**
Allows multiple filters to be joined by the "or" operator, rather than the previously assumed "and".

cc @dchirst

### CODE REVIEW

File: src/osdatahub/FeaturesAPI/features_api.py [hunk 1]
```diff
@@ -46,12 +47,16 @@ class FeaturesAPI:
     }
 
     def __init__(self, key: str, product_name: str, extent: Extent,
-                 spatial_filter_type: str = "intersects"):
+                 spatial_filter_type: str = "intersects", filter_join_type: str = "And"):
+        if filter_join_type not in {"And", "Or"}:
```
Comment (dchirst) [inline src/osdatahub/FeaturesAPI/features_api.py:51]: And and Or could be lowercase to be more pythonic

File: src/osdatahub/FeaturesAPI/features_api.py [hunk 2]
```diff
@@ -46,12 +47,16 @@ class FeaturesAPI:
...
+            raise ValueError("Invalid filter join type '%s' (must be 'And' or 'Or')" % filter_join_type)
+
         self.key = key
         self.product = product_name
         self.extent = extent
         self.filters = []
         self.__spatial_filter = SpatialFilterTypes.get(spatial_filter_type)
```
Comment (dchirst) [inline src/osdatahub/FeaturesAPI/features_api.py:58]: We might want to have a more complex filtering system to allow for some mix of "and" and "or". 
For example, if we have 3 filters (A, B, C), we want to allow A && (B || C), or any other variants. 
Could even use the new Union | operator for python 3.10?

**Review by dchirst — CHANGES_REQUESTED** [review]
Great work! We could make this more complicated if we wanted - question is, do we want to?

**Comment by E-Paine** [comment]
The checks appear to be failing because the API key is empty. iirc, Github doesn't allow access to secrets on the branches of third parties (for security).

File: src/osdatahub/FeaturesAPI/features_api.py (hunk 2 above)
Comment (E-Paine) [reply src/osdatahub/FeaturesAPI/features_api.py:58]: I don't think I makes sense to implement the union operator in this case. However, I agree that the logical and/or operators would be useful to allow a more complicated combination of filters.

File: src/osdatahub/FeaturesAPI/features_api.py (hunk 1 above)
Comment (JEPooleyOS) [reply src/osdatahub/FeaturesAPI/features_api.py:51]: Think I agree here, though obviously you'll need to use the capitalised versions in the actual filter. Maybe the spatial filter could be a property with a getter & setter? Also, out of interest, is there a reason to construct a set to check the value in line 51, as opposed to a tuple or list?

Comment (E-Paine) [reply src/osdatahub/FeaturesAPI/features_api.py]: N/A due as latest commit removes the argument. Marking as resolved.

File: src/osdatahub/FeaturesAPI/features_api.py [hunk 3]
```diff
@@ -102,7 +98,7 @@ def query(self, limit: int = 100) -> FeatureCollection:
             while n_required > 0 and data.grown:
                 params.update({"count": n_required, "startIndex": len(data)})
                 response = requests.get(self.ENDPOINT, params=params)
-                if is_new_api(response):
+                if is_new_api(response.json()):
```
Comment (E-Paine) [inline src/osdatahub/FeaturesAPI/features_api.py:101]: Fixes an issue I missed in #33

File: src/osdatahub/filters.py [hunk 4]
```diff
@@ -1,21 +1,79 @@
...
+        Filter: A valid OGC XML filter
+    """
+    return filter1 | filter2
+
+
+def filter_and(filter1: Filter, filter2: Filter) -> Filter:
+    """Constructs an OGC XML filter that performs an 'and' on the two give filters
+
+    Args:
+        filter1 (Filter):
```
Comment (dchirst) [inline src/osdatahub/filters.py:55]: Probably should add a small  description in these docstrings

File: src/osdatahub/filters.py [hunk 5]
```diff
@@ -1,21 +1,79 @@
...
+        filter2 (Filter):
+
+    Returns:
+        Filter: A valid OGC XML filter
+    """
+    return filter1 | filter2
+
+
+def filter_and(filter1: Filter, filter2: Filter) -> Filter:
```
Comment (dchirst) [inline src/osdatahub/filters.py:51]: I think this could take an arbitrary number of filters, and perform an and on all of them. Same with filter_or()

**Review by dchirst — CHANGES_REQUESTED** [review]
Only a couple of tiny things with the filter_or() and filter_and() functions

File: src/osdatahub/filters.py [hunk 6]
```diff
@@ -1,21 +1,79 @@
...
+        filters (Filter): The filters to be joined
+
+    Returns:
+        Filter: A valid OGC XML filter
+    """
+    return functools.reduce(operator.or_, filters)
+
+
+def filter_and(*filters: Filter) -> Filter:
+    """Constructs an OGC XML filter that performs an 'and' on the two give filters
```
Comment (JEPooleyOS) [inline src/osdatahub/filters.py:53]: typo: give -> given

File: src/osdatahub/filters.py [hunk 7]
```diff
@@ -1,21 +1,79 @@
...
+
+    def __repr__(self) -> str:
+        return f"<Filter object with xml {self.xml!r}>"
+
+    def __str__(self) -> str:
+        return self.xml
+
+
+def filter_or(*filters: Filter) -> Filter:
+    """Constructs an OGC XML filter that performs an 'or' on the two give filters
```
Comment (JEPooleyOS) [inline src/osdatahub/filters.py:41]: typo: give -> given

**Review by JEPooleyOS — CHANGES_REQUESTED** [review]
Looks awesome :)! Wondering if we could get a few tests in to check it's all working as expected?

**Review by JEPooleyOS — APPROVED** [review]

**Pull request merged**